app/src/features/Batches/BatchResourceDialog.tsx

```diff
@@ -0,0 +1,248 @@
+  open,
+  onClose,
+  setAlert,
+}: BatchResourceDialogType): JSX.Element => {
+  const { t } = useTranslation();
+  const classes = useStyles();
+  const { sourceId: id } = useParams<{ sourceId: string }>();
+
+  const { data: resources } = useApiResourcesListQuery(
+    { source: id },
+    { skip: !Boolean(id) }
+  );
+  const [apiBatchCreate] = useApiBatchesCreateMutation();
+
+  const [selectedResourceIds, setSelectedResourceIds] = useState<string[]>([]);
+  const [displayedResources, setDisplayedResources] = useState(resources);
+
+  const displayedResourcesIds = useMemo(
+    () => displayedResources?.map(({ id }) => id),
+    [displayedResources]
+  );
+
+  const areAllDisplayedResourcesSelected = useMemo(() => {
+    if (displayedResourcesIds) {
+      const selectedAndDisplayedResourcesId = selectedResourceIds.filter((id) =>
+        displayedResourcesIds.includes(id)
+      );
+      return (
+        displayedResourcesIds.length > 0 &&
+        selectedAndDisplayedResourcesId.length === displayedResourcesIds.length
+      );
+    }
+  }, [selectedResourceIds, displayedResourcesIds]);
+
+  const isCheckboxIndeterminate = useMemo(() => {
+    if (displayedResourcesIds) {
+      const selectedAndDisplayedResourcesId = selectedResourceIds.filter((id) =>
+        displayedResourcesIds.includes(id)
+      );
+      return (
+        selectedAndDisplayedResourcesId.length > 0 &&
+        selectedAndDisplayedResourcesId.length < displayedResourcesIds.length
+      );
+    }
+  }, [selectedResourceIds, displayedResourcesIds]);
+
+  const onCloseModal = () => {
+    setSelectedResourceIds([]);
+    onClose(false);
+  };
+
+  const handleSearchResource = (
+    e: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
+  ) => {
+    const searchValue = e.target.value.toLowerCase();
+    setDisplayedResources(
+      resources?.filter(
+        ({ label, definition_id }) =>
+          label?.toLowerCase().includes(searchValue) ||
+          definition_id.toLowerCase().includes(searchValue)
+      )
+    );
+  };
+
+  const handleSelectResources = (id: string) => {
+    if (selectedResourceIds.includes(id)) {
+      setSelectedResourceIds(
+        selectedResourceIds.filter((selectedId) => selectedId !== id)
+      );
+    } else {
+      setSelectedResourceIds([...selectedResourceIds, id]);
+    }
+  };
+
+  const handleSelectAllClick = () => {
+    if (displayedResourcesIds && selectedResourceIds.length === 0) {
+      setSelectedResourceIds(displayedResourcesIds);
+    } else if (areAllDisplayedResourcesSelected || isCheckboxIndeterminate) {
+      const resourcesIdsToKeep = selectedResourceIds.filter(
+        (id) => !displayedResourcesIds?.includes(id)
+      );
+      setSelectedResourceIds(resourcesIdsToKeep);
+    } else if (displayedResourcesIds) {
+      const newSelectedResourcesIds = selectedResourceIds.concat(
+        displayedResourcesIds
+      );
+      setSelectedResourceIds(newSelectedResourcesIds);
+    }
```

Comment:
You can actually merge the 1st and the 3rd if statements. The 1st one is only a particular case of the 3rd one, more general.

```suggestion
    if (areAllDisplayedResourcesSelected || isCheckboxIndeterminate) {
      const resourcesIdsToKeep = selectedResourceIds.filter(
        (id) => !displayedResourcesIds?.includes(id)
      );
      setSelectedResourceIds(resourcesIdsToKeep);
    } else if (displayedResourcesIds) {
      const newSelectedResourcesIds = selectedResourceIds.concat(
        displayedResourcesIds
      );
      setSelectedResourceIds(newSelectedResourcesIds);
    }
```